### PR TITLE
UI tests for portal

### PR DESCRIPTION
**Pull request by ellis-johnson**
### Which issue this PR addresses:

Fixes https://msazure.visualstudio.com/AzureRedHatOpenShift/_workitems/edit/12864096 

### What this PR does / why we need it:

Implements some basic initial UI Tests for the admin portal

### Test plan for issue:

These are tests that run in the E2E pipeline

### Is there any documentation that needs to be updated for this PR?

Internal ADO documentation on how to run tests

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -344,6 +345,5 @@ replace (
 	sigs.k8s.io/cluster-api-provider-openstack => github.com/openshift/cluster-api-provider-openstack v0.0.0-20210820223719-a7442bb18bce
 	sigs.k8s.io/kustomize/kyaml => sigs.k8s.io/kustomize/kyaml v0.13.0
 	sigs.k8s.io/structured-merge-diff => sigs.k8s.io/structured-merge-diff v1.0.1-0.20191108220359-b1b620dd3f06
-	sourcegraph.com/sourcegraph/go-diff => github.com/sourcegraph/go-diff v0.5.1
```
Comment (petrkotas) [inline go.mod:347]: why are these removed?

File: hack/portalauth/portalauth.go [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+	}
+
+	portalKeyvaultURI, err := keyvault.URI(_env, env.PortalKeyvaultSuffix)
+	if err != nil {
+		return err
+	}
+
+	portalKeyvault := keyvault.NewManager(msiKVAuthorizer, portalKeyvaultURI)
+
+	sessionKey, err := portalKeyvault.GetBase64Secret(ctx, env.PortalServerSessionKeySecretName, "")
```
Comment (petrkotas) [inline hack/portalauth/portalauth.go:56]: Is this the same key as used in production?

File: test/e2e/setup.go [hunk 3]
```diff
@@ -196,6 +267,15 @@ func setup(ctx context.Context) error {
 		return err
 	}
 
+	log.Infof("Starting Selenium Grid")
```
Comment (petrkotas) [inline test/e2e/setup.go:270]: How costly are the UI tests? I would love to see them hidden behind a flag, so we can run them when required.

**Review by petrkotas — CHANGES_REQUESTED** [review]
Hi @ellis-johnson thank you for the continuous effort in improving our portal. I have a few questions about the implementation, which I believe could add additional value to this PR.
Thank you.

File: test/e2e/setup.go (hunk 3 above)
Comment (ellis-johnson) [reply test/e2e/setup.go:270]: In terms of resources or time? Resources wise I've had to run the portal as part of the E2E suite. And time wise they don't take any longer than a minute or so. But we can still hide them behind a flag that defaults to on if need be

File: hack/portalauth/portalauth.go (hunk 2 above)
Comment (ellis-johnson) [reply hack/portalauth/portalauth.go:56]: I don't believe so, but these tests have been written to only run in development currently. We would need to figure out a way to run this in production

File: go.mod (hunk 1 above)
Comment (ellis-johnson) [reply go.mod:347]: I have no idea, I ran go mod vendor cause of the new packages I added. I'll do some digging to figure out why this has gone missing

Comment (ellis-johnson) [reply go.mod]: Could not find a good reason these were removed. Added them back in, ran go mod vendor and go mod tidy and committed the changes

File: test/e2e/admin_portal.go [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+		text, err := wd.FindElement(ByCSSSelector, "span.itemsCount-162")
+		if err != nil {
+			panic(err)
+		}
+
+		Expect(text.Text()).To(Equal("Showing 0 items"))
+	})
+
+	It("Should be able to populate cluster info panel correctly", func() {
+		testValues := [17]string{
```
Comment (hawkowl) [inline test/e2e/admin_portal.go:64]: where do these test values come from?

Comment (ellis-johnson) [reply test/e2e/admin_portal.go:64]: Ooh, these are just the random hard coded values from my local environment. I definitely missed going back to fix them because the test seems to pass on the CT... Definitely something not quite right there...

Comment (ellis-johnson) [reply test/e2e/admin_portal.go]: I have modified these tests, the css selector was wrong and returning nothing, so these should work now

**Review by petrkotas — CHANGES_REQUESTED** [review]
I should have looked at the commits first. My apologies.
@ellis-johnson nothing changes to what I already said. I have one last ask. Would you please tidy up commit history to stream lined fashion. After lets merge it.

File: test/e2e/admin_portal.go [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+			if err != nil {
+				panic(err)
+			}
+
+			if panelText != ":" {
+				filteredPanelFields = append(filteredPanelFields, panelText)
+			}
+		}
+
+		Expect(panelFields).ShouldNot(BeEmpty())
```
Comment (s-amann) [inline test/e2e/admin_portal.go:91]: This expectation failed in the pipeline: https://msazure.visualstudio.com/AzureRedHatOpenShift/_build/results?buildId=56020765&view=logs&j=3965c7ea-79de-5618-9d9a-42b627bfb354&t=4810cc2a-45e4-5107-9a11-ae0073a144ef&l=1431

Comment (ellis-johnson) [reply test/e2e/admin_portal.go]: Finally resolved!

**Comment by facchettos** [comment]
Just a comment, this branch is quite dirty with commits from other PRs. That would be great if we could get a cleaner history (and also clean up some weird commits (`I'm so annoyed and stupid`, for example :D ) before we merge it

File: .gitignore [hunk 6]
```diff
@@ -30,3 +30,5 @@ gomock_reflect_*
 /portal/v1/node_modules/
 /portal/v2/node_modules/
 .idea*
+/hack/hive-config/crds
```
[the remark this answers is not in the document]
Comment (ellis-johnson) [reply .gitignore:33]: I believe this was from a conflict where I had updated the gitignore

**Comment by ellis-johnson** [comment]
> Just a comment, this branch is quite dirty with commits from other PRs. That would be great if we could get a cleaner history (and also clean up some weird commits (`I'm so annoyed and stupid`, for example :D ) before we merge it

Have cleaned up commits and removed the silly ones. Let me know if theres anything else to change

File: test/e2e/setup.go [hunk 7]
```diff
@@ -73,6 +83,156 @@ func skipIfNotInDevelopmentEnv() {
 	}
 }
 
+func TakeScreenshot(wd WebDriver, e error) {
+	log.Info("Taking Screenshot and saving page source")
```
Comment (s-amann) [inline test/e2e/setup.go:87]: this is cool, do you have any example runs with the screenshots still lingering around to look at?

**Review by s-amann — APPROVED** [review]
LGTM

File: test/e2e/setup.go [hunk 8]
```diff
@@ -73,6 +83,156 @@ func skipIfNotInDevelopmentEnv() {
...
+		panic(err)
+	}
+
+	log := utillog.GetLogger()
+
+	gob.Register(time.Time{})
+
+	// Navigate to the simple playground interface.
+	host, isNotLocal := os.LookupEnv("PORTAL_HOSTNAME")
+	if !isNotLocal {
```
Comment (bennerv) [inline test/e2e/setup.go:196]: Double negatives are hard.  Essentially what this code is doing (to my understanding) is if a portal hostname is set (in the case of a remote portal) use the env var.  Otherwise default to localhost.  I would do something like this, or sub out `/s/exists/ok` as per normal fashion.  
```suggestion
	host, exists := os.LookupEnv("PORTAL_HOSTNAME")
	if !exists {
```

File: test/e2e/setup.go [hunk 9]
```diff
@@ -73,6 +83,156 @@ func skipIfNotInDevelopmentEnv() {
...
+	}
+	return host, &wd
+}
+
+func adminPortalSessionTearDown() {
+	log.Infof("Stopping Selenium Grid")
+	cmd := exec.Command("docker", "rm", "--force", "selenium-edge-standalone")
+
+	output, err := cmd.CombinedOutput()
+
```
Comment (bennerv) [inline test/e2e/setup.go:230]: nit - remove empty new line here

File: test/e2e/setup.go [hunk 10]
```diff
@@ -153,6 +313,31 @@ func newClientSet(ctx context.Context) (*clientSet, error) {
...
+	log.Infof("Starting Selenium Grid")
+	cmd := exec.CommandContext(ctx, "docker", "pull", "selenium/standalone-edge:latest")
+
+	output, err := cmd.CombinedOutput()
+
+	log.Infof("Selenium Image Pull Output : %s\n", output)
+
+	if err != nil {
+		log.Fatalf("Error occurred pulling selenium image\n Output: %s\n Error: %s\n", output, err)
+	}
```
Comment (bennerv) [inline test/e2e/setup.go:326]: Let's move this around
```suggestion
	output, err := cmd.CombinedOutput()
	if err != nil {
		log.Fatalf("Error occurred pulling selenium image\n Output: %s\n Error: %s\n", output, err)
	}

	log.Infof("Selenium Image Pull Output : %s\n", output)

File: test/e2e/setup.go [hunk 11]
```diff
@@ -153,6 +313,31 @@ func newClientSet(ctx context.Context) (*clientSet, error) {
...
+
+	output, err = cmd.CombinedOutput()
+
+	log.Infof("Selenium Container Run Output : %s\n", output)
+
+	if err != nil {
+		log.Fatalf("Error occurred starting selenium grid\n Output: %s\n Error: %s\n", output, err)
+	}
+
+	return err
```
Comment (bennerv) [inline test/e2e/setup.go:338]: refactor:
```suggestion
	output, err = cmd.CombinedOutput()
	if err != nil {
		log.Fatalf("Error occurred starting selenium grid\n Output: %s\n Error: %s\n", output, err)
	}

	log.Infof("Selenium Container Run Output : %s\n", output)
	return nil

**Review by bennerv — CHANGES_REQUESTED** [review]
first pass - few nits but great work so far

File: test/e2e/setup.go [hunk 12]
```diff
@@ -73,6 +83,156 @@ func skipIfNotInDevelopmentEnv() {
...
+
+	panic(e)
+}
+
+func adminPortalSessionSetup() (string, *WebDriver) {
+	const (
+		port = 4444
+	)
+
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
```
[the remark this answers is not in the document]
Comment (ellis-johnson) [reply test/e2e/setup.go:158]: I believe originally I put this in as some kind of fail safe for the New Selenium Remote hanging but after running it multiple times it has not done it again, so I will remove it for now and see how it performs. If the timeout is needed again I will use a better method.

File: test/e2e/setup.go (hunk 8 above)
Comment (ellis-johnson) [reply test/e2e/setup.go:196]: Fixed!

File: test/e2e/setup.go (hunk 9 above)
Comment (ellis-johnson) [reply test/e2e/setup.go:230]: Fixed!

File: test/e2e/setup.go (hunk 10 above)
Comment (ellis-johnson) [reply test/e2e/setup.go:326]: Sufficiently shuffled

File: test/e2e/setup.go (hunk 11 above)
Comment (ellis-johnson) [reply test/e2e/setup.go]: Refactored

File: pkg/cluster/install.go [hunk 13]
```diff
@@ -42,6 +42,7 @@ func (m *manager) adminUpdate() []steps.Step {
 	// don't require a running cluster
 	toRun := []steps.Step{
 		steps.Action(m.initializeKubernetesClients), // must be first
+		steps.Action(m.initializeOperatorDeployer),  // depends on kube clients
```
[the remark this answers is not in the document]
Comment (ellis-johnson) [reply pkg/cluster/install.go]: Yeah seems like a side effect of the dirty rebasing I was doing, have fixed

File: test/e2e/setup.go [hunk 14]
```diff
@@ -73,6 +83,156 @@ func skipIfNotInDevelopmentEnv() {
 	}
 }
 
+func TakeScreenshot(wd WebDriver, e error) {
```
[the remark this answers is not in the document]
Comment (ellis-johnson) [reply test/e2e/setup.go:86]: Yep good call, have fixed

File: test/e2e/setup.go (hunk 7 above)
Comment (ellis-johnson) [reply test/e2e/setup.go:88]: The builds have expired unfortunately... But in the APAC Demos I did show off what it looks like in ADO if you wanted to see that https://drive.google.com/file/d/1wvdk6CfWxY7K_9rNvhH90lyNg1dURH9-/view?usp=sharing 

It starts at 25:10

Comment (s-amann) [reply test/e2e/setup.go:88]: awesome, thank you!

**Review by s-amann — APPROVED** [review]

File: pkg/portal/bindata.go [hunk 15]
```diff
@@ -10,6 +10,8 @@
 // ../../portal/v2/build/static/js/2.fc3140a1.chunk.js
```
Comment (bennerv) [inline pkg/portal/bindata.go:10]: @facchettos would ask you to remove bindata from these.  Can we do so in a follow-up PR though? 

See #2119 as an example.

File: rebase.sh [hunk 16]
```diff
@@ -0,0 +1,2 @@
+#!/bin/bash
+git checkout --theirs . & git add . & git rebase --continue
```
Comment (bennerv) [inline rebase.sh:2]: seems like this should be a hack script.  Don't think it should live top-level.

File: test/e2e/setup.go [hunk 17]
```diff
@@ -73,6 +83,151 @@ func skipIfNotInDevelopmentEnv() {
...
+
+	caps := Capabilities{
+		"browserName":         "MicrosoftEdge",
+		"acceptInsecureCerts": true,
+	}
+	wd := WebDriver(nil)
+
+	var err error
+
+	time.Sleep(time.Second * 10)
```
Comment (bennerv) [inline test/e2e/setup.go:169]: Why sleep here?  Should we try waiting on something instead?

File: test/e2e/setup.go [hunk 18]
```diff
@@ -73,6 +83,151 @@ func skipIfNotInDevelopmentEnv() {
...
+
+func adminPortalSessionSetup() (string, *WebDriver) {
+	const (
+		port = 4444
+	)
+
+	os.Setenv("SE_SESSION_REQUEST_TIMEOUT", "9000")
+
+	caps := Capabilities{
+		"browserName":         "MicrosoftEdge",
```
Comment (bennerv) [inline test/e2e/setup.go:162]: `MicrosoftEdge`
:upside_down_face: :wink:

**Review by bennerv — CHANGES_REQUESTED** [review]
few more questions - hope this is my last pass at it :)

File: pkg/portal/bindata.go (hunk 15 above)
Comment (ellis-johnson) [reply pkg/portal/bindata.go:10]: Yep can do, will create an ADO ticket to do this

File: rebase.sh (hunk 16 above)
Comment (ellis-johnson) [reply rebase.sh:2]: This was accidentally committed. Will remove

File: test/e2e/setup.go (hunk 17 above)
Comment (ellis-johnson) [reply test/e2e/setup.go:169]: I think this was a debug sleep that was left in. Will remove

File: test/e2e/setup.go [hunk 19]
```diff
@@ -73,6 +83,151 @@ func skipIfNotInDevelopmentEnv() {
...
+		"acceptInsecureCerts": true,
+	}
+	wd := WebDriver(nil)
+
+	var err error
+
+	_, err = url.ParseRequestURI("https://localhost:4444")
```
Comment (facchettos) [inline test/e2e/setup.go:169]: Nitpicking: we could just initialize it in one go. If you have some more commits to push , it would be great to include it, otherwise I think we can let it pass

**Review by s-amann — DISMISSED** [review]

File: test/e2e/setup.go (hunk 19 above)
Comment (ellis-johnson) [reply test/e2e/setup.go]: Fixed!

File: test/e2e/admin_portal.go [hunk 20]
```diff
@@ -0,0 +1,215 @@
...
+		wd.Wait(ElementIsLocated(ByCSSSelector, "button[aria-label='SSH']"))
+
+		button, err := wd.FindElement(ByCSSSelector, "button[aria-label='SSH']")
+		if err != nil {
+			TakeScreenshot(wd, err)
+		}
+
+		button.Click()
+
+		wd.Wait(ElementIsLocated(ByID, "ModalFocusTrapZone25"))
```
Comment (hawkowl) [inline test/e2e/admin_portal.go:176]: By defining a trapzoneprops and passing it to the modal, we can set this to what we want:

```
  const trapzoneprops: IFocusTrapZoneProps = {id: "mytrapzoneId"};

...
<Modal
focusTrapZoneProps={trapzoneprops}
```

Since this modal is only used once and isn't unique for each cluster, I don't think we need to do more than have a good ID.

File: test/e2e/admin_portal.go [hunk 21]
```diff
@@ -0,0 +1,215 @@
...
+
+		wd.Wait(ElementIsLocated(ByID, "Dropdown55"))
```
Comment (hawkowl) [inline test/e2e/admin_portal.go:178]: For the dropdown, we can pass `id` in to the props directly to set it:

```
<Dropdown
id="foo" 
```

Since there's only one SSH modal ,I don't think we need to randomise that.

File: test/e2e/admin_portal.go [hunk 22]
```diff
@@ -0,0 +1,215 @@
...
+
+		wd.Wait(ElementIsLocated(ByID, "Dropdown55-list0"))
+		machine, err := wd.FindElement(ByID, "Dropdown55-list0")
+		if err != nil {
+			TakeScreenshot(wd, err)
+		}
+
+		machine.Click()
+
+		wd.Wait(ElementIsLocated(ByID, "id__56"))
```
Comment (hawkowl) [inline test/e2e/admin_portal.go:194]: We can set an `id` on the PrimaryButton in the modal to give this something fixed.

File: test/e2e/admin_portal.go [hunk 23]
```diff
@@ -0,0 +1,215 @@
...
+		if err != nil {
+			TakeScreenshot(wd, err)
+		}
+
+		requestBtn.Click()
+
+		wd.Wait(ElementIsLocated(ByID, "title24"))
+
+		// Test fails if these labels aren't present
+		err = wd.Wait(ElementIsLocated(ByID, "TextFieldLabel72"))
```
Comment (hawkowl) [inline test/e2e/admin_portal.go:205]: We can also set id on these text field labels to give them fixed IDs.

**Review by s-amann — APPROVED** [review]

**Review by s-amann — DISMISSED** [review]

**Review by s-amann — DISMISSED** [review]

File: hack/e2e/run-rp-and-e2e.sh [hunk 24]
```diff
@@ -37,6 +37,42 @@ kill_rp(){
     wait $rppid
 }
 
+validate_portal_running() {
+    echo "########## ？Checking Admin Portal Status ##########"
+    ELAPSED=0
+    while true; do
+        http_code=$(curl -k -s -o /dev/null -w '%{http_code}' https://localhost:8444/api/info || true)
```
Comment (darthhexx) [inline hack/e2e/run-rp-and-e2e.sh:44]: The ` || true` is very javascripty in this context :) and not required.

File: hack/portalauth/portalauth.go [hunk 25]
```diff
@@ -0,0 +1,96 @@
...
+	session := sessions.NewSession(store, SessionName)
+	opts := *store.Options
+	session.Options = &opts
+
+	session.Values[SessionKeyUsername] = username
+	session.Values[SessionKeyGroups] = strings.Split(*groups, ",")
+	session.Values[SessionKeyExpires] = time.Now().Add(time.Hour)
+
+	encoded, err := securecookie.EncodeMulti(session.Name(), session.Values,
+		store.Codecs...)
```
Comment (darthhexx) [inline hack/portalauth/portalauth.go:77]: This shouldn't be on a new line.

File: hack/portalauth/portalauth.go [hunk 26]
```diff
@@ -0,0 +1,96 @@
...
+	if err != nil {
+		return err
+	}
+
+	// encoded
```
Comment (darthhexx) [inline hack/portalauth/portalauth.go:82]: No need for this comment.

File: test/e2e/admin_portal.go [hunk 27]
```diff
@@ -0,0 +1,198 @@
...
+	. "github.com/serge1peshcoff/selenium-go-conditions" //nolint
+	. "github.com/tebeka/selenium"                       //nolint
+)
+
+var _ = Describe("Admin Portal E2E Testing", func() {
+	BeforeEach(skipIfNotInDevelopmentEnv)
+	var wdPoint *WebDriver
+	var wd WebDriver
+	var host string
+	JustBeforeEach(func() {
```
Comment (darthhexx) [inline test/e2e/admin_portal.go:21]: Should be a blank line above this.

File: test/e2e/admin_portal.go [hunk 28]
```diff
@@ -0,0 +1,198 @@
...
+		wd.Get(host + "/v2")
+		wd.Refresh()
+	})
+
+	AfterEach(func() {
+		if wd != nil {
+			wd.Quit()
+		}
+	})
+	It("Should be able to populate cluster data correctly", func() {
```
Comment (darthhexx) [inline test/e2e/admin_portal.go:33]: Should be a blank line above this.

File: test/e2e/admin_portal.go [hunk 29]
```diff
@@ -0,0 +1,198 @@
...
+		if err != nil {
+			TakeScreenshot(wd, err)
+		}
+
+		Expect(text.Text()).To(Equal("Showing 0 items"))
+	})
+
+	It("Should be able to populate cluster info panel correctly", func() {
+		const CLUSTER_INFO_HEADINGS = 17
+		err := wd.Wait(ElementIsLocated(ByCSSSelector, "div[data-automation-key='name']"))
```
Comment (darthhexx) [inline test/e2e/admin_portal.go:69]: Should be an empty line above this, in order to separate the const declaration from the function call.

File: test/e2e/admin_portal.go [hunk 30]
```diff
@@ -0,0 +1,198 @@
...
+		if err != nil {
+			TakeScreenshot(wd, err)
+		}
+
+		// Paste clipboard
+		filter.Click()
+		filter.SendKeys(ControlKey + "v")
+
+		resourceId, err := filter.GetAttribute("value")
+
```
Comment (darthhexx) [inline test/e2e/admin_portal.go:148]: There shouldn't be an empty line between the function call and the error checking.

File: test/e2e/admin_portal.go [hunk 31]
```diff
@@ -0,0 +1,198 @@
...
+		if err != nil {
+			TakeScreenshot(wd, err)
+		}
+
+		button.Click()
+
+		wd.Wait(ElementIsLocated(ByID, "sshModal"))
+
+		wd.Wait(ElementIsLocated(ByID, "sshDropdown"))
+		sshDropdown, err := wd.FindElement(ByID, "sshDropdown")
```
Comment (darthhexx) [inline test/e2e/admin_portal.go:169]: Would read a little easier if the _Wait_'s were grouped separately from the function call and error check:
```go
		wd.Wait(ElementIsLocated(ByID, "sshModal"))
		wd.Wait(ElementIsLocated(ByID, "sshDropdown"))

		sshDropdown, err := wd.FindElement(ByID, "sshDropdown")
		if err != nil {
			TakeScreenshot(wd, err)
		}

File: test/e2e/setup.go [hunk 32]
```diff
@@ -75,6 +85,149 @@ func skipIfNotInDevelopmentEnv() {
...
+
+	os.Setenv("SE_SESSION_REQUEST_TIMEOUT", "9000")
+
+	caps := Capabilities{
+		"browserName":         "MicrosoftEdge",
+		"acceptInsecureCerts": true,
+	}
+	wd := WebDriver(nil)
+
+	_, err := url.ParseRequestURI("https://localhost:4444")
```
Comment (darthhexx) [inline test/e2e/setup.go:169]: The `4444` should be substituted by the `port` const.

File: test/e2e/setup.go [hunk 33]
```diff
@@ -75,6 +85,149 @@ func skipIfNotInDevelopmentEnv() {
...
+	if err != nil {
+		panic(err)
+	}
+
+	for i := 0; i < 10; i++ {
+		wd, err = NewRemote(caps, fmt.Sprintf("http://localhost:%d/wd/hub", port))
+		time.Sleep(time.Second)
```
Comment (darthhexx) [inline test/e2e/setup.go:176]: This Sleep should be after the `wd != nil` check, so that we don't wait for a second when we have what we want, only when we don't.

File: test/e2e/setup.go [hunk 34]
```diff
@@ -75,6 +85,149 @@ func skipIfNotInDevelopmentEnv() {
...
+	}
+
+	log := utillog.GetLogger()
+
+	gob.Register(time.Time{})
+
+	// Navigate to the simple playground interface.
+	host, exists := os.LookupEnv("PORTAL_HOSTNAME")
+	if !exists {
+		host = "https://localhost:8444"
```
Comment (darthhexx) [inline test/e2e/setup.go:194]: This default host value should have the port declared at the top of this function.

File: test/e2e/setup.go [hunk 35]
```diff
@@ -75,6 +85,149 @@ func skipIfNotInDevelopmentEnv() {
...
+	}
+	if err := wd.Get(host + "/api/info"); err != nil {
+		log.Infof("Could not get to %s. With error : %s", host, err.Error())
+	}
+	cmd := exec.Command("go", "run", "./hack/portalauth", "-username", "test", "-groups", "$AZURE_PORTAL_ELEVATED_GROUP_IDS", "2>", "/dev/null")
```
Comment (darthhexx) [inline test/e2e/setup.go:199]: There should be an empty line above this, to separate it from the previous call and error check.

File: test/e2e/setup.go [hunk 36]
```diff
@@ -75,6 +85,149 @@ func skipIfNotInDevelopmentEnv() {
...
+	output, err := cmd.Output()
+
+	if err != nil {
```
Comment (darthhexx) [inline test/e2e/setup.go:202]: For readability, it would be clearer to group the error check with the call above it.

File: test/e2e/admin_portal.go [hunk 37]
```diff
@@ -0,0 +1,198 @@
...
+	AfterEach(func() {
+		if wd != nil {
+			wd.Quit()
+		}
+	})
+	It("Should be able to populate cluster data correctly", func() {
+		err := wd.Wait(ElementIsLocated(ByCSSSelector, "div[data-automation-key='name']"))
+		if err != nil {
+			TakeScreenshot(wd, err)
+		}
```
Comment (darthhexx) [inline test/e2e/admin_portal.go:37]: It is not evident from this code block, and all 20-odd that follow the same pattern, that the `TakeScreenshot` function is going to panic on purpose. IMO we should either change the function name or the way we handle these to be more explicit.

Simplest option is to change the name to something like `SaveScreenshotAndPanic`, but ideally we would change the behaviour to not panic and exit with grace. That's just my personal preference though :)

**Review by darthhexx — CHANGES_REQUESTED** [review]
A few last round nits to address.

File: hack/e2e/run-rp-and-e2e.sh (hunk 24 above)
Comment (ellis-johnson) [reply hack/e2e/run-rp-and-e2e.sh]: Fixed!

File: hack/portalauth/portalauth.go (hunk 26 above)
Comment (ellis-johnson) [reply hack/portalauth/portalauth.go]: Removed

File: test/e2e/admin_portal.go (hunk 28 above)
Comment (ellis-johnson) [reply test/e2e/admin_portal.go:35]: Added

File: test/e2e/admin_portal.go (hunk 30 above)
Comment (ellis-johnson) [reply test/e2e/admin_portal.go:150]: Removed

File: test/e2e/admin_portal.go (hunk 31 above)
Comment (ellis-johnson) [reply test/e2e/admin_portal.go]: Fixed

File: test/e2e/setup.go (hunk 33 above)
Comment (ellis-johnson) [reply test/e2e/setup.go]: Moved

File: test/e2e/setup.go (hunk 35 above)
Comment (ellis-johnson) [reply test/e2e/setup.go:202]: Added

File: hack/portalauth/portalauth.go (hunk 25 above)
Comment (ellis-johnson) [reply hack/portalauth/portalauth.go]: Fixed

File: test/e2e/admin_portal.go (hunk 27 above)
Comment (ellis-johnson) [reply test/e2e/admin_portal.go:22]: Added

File: test/e2e/admin_portal.go (hunk 29 above)
Comment (ellis-johnson) [reply test/e2e/admin_portal.go:72]: Added

File: test/e2e/setup.go (hunk 32 above)
Comment (ellis-johnson) [reply test/e2e/setup.go]: Fixed

File: test/e2e/setup.go (hunk 36 above)
Comment (ellis-johnson) [reply test/e2e/setup.go:204]: Fixed

**Review by darthhexx — CHANGES_REQUESTED** [review]

File: hack/e2e/run-rp-and-e2e.sh [hunk 38]
```diff
@@ -37,12 +37,51 @@ kill_rp(){
...
+kill_portal(){
+    echo "########## Kill the Admin Portal running in background ##########"
+    rppid=$(lsof -t -i :8444)
+    kill $rppid
+    wait $rppid
+}
+
 run_vpn() {
+    echo "########## 🚀 Run OpenVPN in background ##########"
+    echo "Using Secret secrets/$VPN"
```
[the remark this answers is not in the document]
Comment (ellis-johnson) [reply hack/e2e/run-rp-and-e2e.sh:78]: Not for the admin portal but this was a merge conflict from a prior change in master

File: .pipelines/e2e.yml [hunk 39]
```diff
@@ -44,16 +44,32 @@ jobs:
...
       . ./hack/e2e/run-rp-and-e2e.sh
-      trap 'set +e; kill_rp; clean_e2e_db; kill_vpn' EXIT
+      trap 'set +e; kill_rp; kill_portal; clean_e2e_db; kill_vpn' EXIT
 
       run_vpn
 
       deploy_e2e_db
 
+      run_portal
+      sleep 120
```
[the remark this answers is not in the document]
Comment (ellis-johnson) [reply .pipelines/e2e.yml:54]: Yep this was put in for debugging, have removed it

File: hack/portalauth/portalauth.go [hunk 40]
```diff
@@ -0,0 +1,94 @@
...
+	session.Values[SessionKeyUsername] = username
+	session.Values[SessionKeyGroups] = strings.Split(*groups, ",")
+	session.Values[SessionKeyExpires] = time.Now().Add(time.Hour)
+
+	encoded, err := securecookie.EncodeMulti(session.Name(), session.Values, store.Codecs...)
+	if err != nil {
+		return err
+	}
+
+	fmt.Printf("%s", encoded)
```
[the remark this answers is not in the document]
Comment (ellis-johnson) [reply hack/portalauth/portalauth.go:81]: Removed

File: test/e2e/admin_portal.go (hunk 37 above)
Comment (ellis-johnson) [reply test/e2e/admin_portal.go]: Fixed

**Review by s-amann — APPROVED** [review]

**Review by darthhexx — DISMISSED** [review]
Thanks for all the work to get this across the line.

**Comment by darthhexx** [comment]
@bennerv and @petrkotas would you mind taking a look when you have some time? We'd like to get this PR merged so that other admin portal PR's can make use of the existing and also add new tests.

**Review by darthhexx — APPROVED** [review]
lgtm

**Review by darthhexx — APPROVED** [review]
LGTM!

**Comment by facchettos** [comment]
I don't think you should have bypassed the dotimport linting. It makes it harder to detect where we are using selenium. Can you revert it ?
